granulate_utils/metrics/sampler.py

```diff
@@ -2,43 +2,84 @@
 
 import psutil
 
 from granulate_utils.exceptions import MissingExePath
 from granulate_utils.linux.ns import resolve_host_path
 from granulate_utils.linux.process import process_exe, search_for_process
+from granulate_utils.metrics import Collector, MetricsSnapshot, Sample
 from granulate_utils.metrics.mode import SPARK_MESOS_MODE, SPARK_STANDALONE_MODE, SPARK_YARN_MODE
+from granulate_utils.metrics.spark import SparkApplicationMetricsCollector
+from granulate_utils.metrics.yarn import YarnCollector
 
 FIND_CLUSTER_TIMEOUT_SECS = 10 * 60
 
 
-class SparkSampler:
+class Sampler(ABC):
+    @abstractmethod
+    def discover(self) -> bool:
+        """
+        The collector's main loop will use this function to determine what Collector's to enabled, if to enable at all.
+        returns True if we have these configurations, False otherwise
+
+        returns True if discover succeeded, False otherwise
+        """
+
+        pass
+
+    @abstractmethod
+    def snapshot(self) -> Optional[MetricsSnapshot]:
+        """
+        This function will be used in a collector loop.
+        It will take care of all the logic to collect metrics from Spark, without any backend communication.
+
+        returns Optional[MetricsSnapshot].
+        """
+        pass
+
+
+class BigDataSampler(Sampler):
     """
     Spark cluster metrics sampler
     """
 
-    def __init__(self, sample_period: float, storage_dir: str, logger: logging.LoggerAdapter, hostname: str):
+    def __init__(
+        self,
+        logger: logging.LoggerAdapter,
+        hostname: str,
+        master_address: Optional[str],
+        cluster_mode: Optional[str],
+        applications_metrics: Optional[bool] = False,
+    ):
         self._logger = logger
-        self._master_address: Optional[str] = None
-        self._spark_mode: Optional[str] = None
-        self._collection_thread: Optional[Thread] = None
-        self._sample_period = sample_period
-        # not the same instance as GProfiler._stop_event
-        self._stop_event = Event()
-        self._stop_collection = False
-        self._is_running = False
         self._hostname = hostname
-        self._storage_dir = storage_dir
-        if self._storage_dir is not None:
-            assert os.path.exists(self._storage_dir) and os.path.isdir(self._storage_dir)
-        else:
-            self._logger.debug("Output directory is None. Will add metrics to queue")
+        self._applications_metrics = applications_metrics
+        self._spark_samplers: List[Collector] = []
+
+        if (cluster_mode is not None) and (master_address is not None):
+            # No need to guess cluster mode and master address
+            self._cluster_mode = cluster_mode
+            self._master_address = f"http://{master_address}"
+        elif (cluster_mode is None) and (master_address is None):
+            # Guess cluster mode and master address
+            cluster_conf = self._guess_cluster_mode()
```

Comment:
Why are you calling `guess` here and not in `discover` only?

Reply:
https://github.com/Granulate/granulate-utils/pull/138/files#diff-e8dd154b7df986eb768fe56114e039f44a244a6a5d75f464611aa95fcb21866fL73